app/src/main/java/org/lineageos/glimpse/viewmodels/ThumbnailViewModel.kt

```diff
@@ -0,0 +1,94 @@
+/*
+ * SPDX-FileCopyrightText: 2023 The LineageOS Project
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.lineageos.glimpse.viewmodels
+
+import android.app.Application
+import androidx.lifecycle.viewModelScope
+import androidx.lifecycle.viewmodel.initializer
+import androidx.lifecycle.viewmodel.viewModelFactory
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.stateIn
+import org.lineageos.glimpse.models.Media
+import org.lineageos.glimpse.recyclerview.ThumbnailAdapter
+import org.lineageos.glimpse.repository.MediaRepository
+import org.lineageos.glimpse.utils.MediaStoreBuckets
+import java.time.ZoneId
+import java.time.temporal.ChronoUnit
+import java.util.Date
+
+class ThumbnailViewModel(
+    application: Application,
+    bucketId: Int,
+    addHeaders: Boolean,
+) : GlimpseViewModel(application) {
+    val mediaWithHeaders = MediaRepository.media(context, bucketId).map { medias ->
+        val data = when (addHeaders) {
+            true -> mutableListOf<DataType>().apply {
+                for (i in medias.indices) {
+                    val currentMedia = medias[i]
+
+                    if (i == 0) {
+                        // First element must always be a header
+                        add(DataType.DateHeader(currentMedia.dateAdded))
+                        add(DataType.Thumbnail(currentMedia))
+                        continue
+                    }
+
+                    val previousMedia = medias[i - 1]
+
+                    val before = previousMedia.dateAdded.toInstant().atZone(ZoneId.systemDefault())
+                    val after = currentMedia.dateAdded.toInstant().atZone(ZoneId.systemDefault())
+                    val days = ChronoUnit.DAYS.between(after, before)
+
+                    if (days >= 1 || before.dayOfMonth != after.dayOfMonth) {
+                        add(DataType.DateHeader(currentMedia.dateAdded))
+                    }
+
+                    add(DataType.Thumbnail(currentMedia))
+                }
+            }
+
+            false -> medias.map {
+                DataType.Thumbnail(it)
+            }
+        }
+
+        QueryResult.Data(data)
+    }.stateIn(
+        viewModelScope,
+        started = SharingStarted.WhileSubscribed(),
+        initialValue = QueryResult.Empty(),
+    )
+
+    sealed class DataType(val id: Long, val viewType: Int) {
+        class Thumbnail(val media: Media) : DataType(media.id, ThumbnailAdapter.ViewTypes.THUMBNAIL.ordinal) {
+            override fun equals(other: Any?) = media == other
+            override fun hashCode() = media.hashCode()
+        }
+
+        class DateHeader(val date: Date) : DataType(-date.time, ThumbnailAdapter.ViewTypes.DATE_HEADER.ordinal) {
```

Comment:
This may not work, Java doesn't have unsigned types so mediaID can be negative ( and is negative )
We need to come up with something unique or use diff util rather than stable ids

Reply:
Done